[SECURITY] JWT 인증/인가 구현 + CORS - #2

hongjunpyo123 opened the pull request:
## 🔥*Pull requests*

#️⃣️ **작업한 브랜치**
- feature/#2

📄  **작업한 내용**
- jwt 생성 및 유효성 검증 기능을 구현하였고 커스텀 필터를 등록하였습니다
- 스웨거 접속에 필요한 엔드포인트를 허용하였습니다
- 토큰이 유효하지 않거나 인가 실패 시 공통 응답 객체를 반환하도록 하였습니다
- 모든 출처를 허용하도록 CORS 를 설정하였습니다

## 🚨 참고 사항
- jwt 유효성 검증 및 생성 은 JWTutil 클래스에서 처리됩니다
- 만료일자는 security.properties 에 명시해놓았지만 토큰 생성 시 직접 지정할 수 있습니다
- 인증 실패 시 응답 객체 반환이 안되는 이슈가 있는데 해결을 못했습니다
- userId 를 원래 Long 타입으로 받았는데 토큰 생성이 안돼서 String으로 변경하였습니다

## 💻 관련 이슈
- Resolved: #2

## 💬 리뷰 요구사항
- (JWTutil) 토큰에 포함되어야 할 유저 정보가  적절한지 확인 부탁드립니다! 
- (JwtAuthenticationFilter) 필터의 인증 흐름과 실패시 반환되는 응답이 적절한지 확인 부탁드립니다!

hongjunpyo123 committed:
chore: jwt, security 의존성 추가

hongjunpyo123 committed:
feat: jwt관련 클래스 파일 추가
- SecurityConfig, JwtAuthenticationFilter, JWTutil

hongjunpyo123 committed:
chore: security.properties 추가
- jwt 키 설정
- TTL 1일로 설정
- 헤더, 접두사 지정

hongjunpyo123 committed:
feat: JWTutil 구현
- getUserId 메서드 추가
- getRole 메서드 추가
- isExpired 메서드 추가
- createToken 메서드 추가
- 클레임: iss, sub, iat, exp, roles

hongjunpyo123 committed:
refactor: createToken 메서드 수정
- 토큰 생성 시 roles 클레임 ROLE_ 접두사 추가되도록 수정

hongjunpyo123 committed:
merge develop branch

hongjunpyo123 committed:
feat: JwtAuthenticationFilter 구현
- 인증 실패 시 json 형식 반환(응답 형식 정해지면 추가 예정)
- 토큰 검증 및 파싱 로직 구현
- SecurityConfig: JwtAuthenticationFilter 추가

hongjunpyo123 committed:
refactor: AuthenticationConfiguration 중복 주입 제거

hongjunpyo123 committed:
refactor: 파일 구조 변경
- config, filter -> api 모듈 안으로 이동
- util 패키지는 common에 유지
- resources -> api모듈로 이동

hongjunpyo123 committed:
feat: 인증/인가 예외 처리 공통 응답 객체 적용

hongjunpyo123 committed:
feat: CORS 설정 추가
- WebConfig 파일 추가
- CORS 설정(프론트 도메인 나오기 전 까지 일단 모두 허용했습니다)

hongjunpyo123 committed:
fix: 버그 수정 및 토큰 발급 테스트
- userId 타입 변경 Long -> String (Long 타입 에러 발생..)
- JwtAuthenticationFilter 에서 설정 파일이 로드가 안됨.. 헤더부분과 접두사 부분 직접 명시 하는 것으로 변경
- 인증/인가 jwt 발급 테스트를 위한 TestController 파일 생성
- 인증 실패시 실패 메시지가 반환되지 않는 문제 있음

Seung-9 commented:
고생하셨습니다👍🏻
근데 혹시 JwtUtil을 api모듈이 아닌 common 모듈이 아닌 들어간 이유가 있을까요?

Seung-9 commented on cubco-common/src/main/java/org/cubco/util/JWTutil.java:
```diff
@@ -1,4 +1,46 @@
+        return Jwts.builder()
+                .claim("iss", "cubco")
+                .claim("sub", userId)
+                .issuedAt(new Date(System.currentTimeMillis()))
+                .expiration(new Date(System.currentTimeMillis() + expiredms))
```
new Date() 부분은 따로 메서드로 분리하는 게 좋을 것 같아요

Seung-9 replied:
그리고 expiredms는 매개변수로 받아오지 말고 환경변수로 받아와서 설정해주면 더 깔끔할 거 같아요

Seung-9 commented on cubco-common/src/main/java/org/cubco/util/JWTutil.java:
```diff
@@ -32,7 +32,7 @@ public Boolean isExpired(String token){
+    public String createToken(String userId, String role, Long expiredms){
         return Jwts.builder()
                 .claim("iss", "cubco")
                 .claim("sub", userId)
```
userId는 그냥 long타입으로 두고 String.valueOf(userId)로 넣으면 됩니다
```java
.claim("sub", userId) -> .setSubject(String.valueOf(userId))
```
그리고 claim에 iss를 넣은 이유가 따로 있을까요 ?

hongjunpyo123 replied:
넵 수정하겠습니다! jwt표준 찾아보면서 만들다보니 넣게됐는데 따로 이유는 없습니다..

Seung-9 commented on cubco-common/src/main/java/org/cubco/util/JWTutil.java:
```diff
@@ -1,4 +1,46 @@
+
+    public Boolean isExpired(String token){
+        return Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token).getPayload().getExpiration().before(new Date());
+    }
```
Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token) 로직이 반복되고 있는데
공통되는 파싱 및 검증 로직을 한 번에 처리하면 가독성이나 유지보수 측면에서 좋아보입니다. 그리고 유효성 검증 단계에서 예외처리가 부족해서 ExpiredJwtException, UnsupportedJwtException, MalformedJwtException, SignatureException, IllegalArgumentException, Exception 등 다양한 상황에 대한 예외처리도 고려해주시면 안정성 측면에서도 훨씬 좋아질 것 같아요

파싱, 검증 및 예외처리 로직을 하나의 메서드에서 구현하고  반환 타입을 Jws<Claims> 로 해주고, JwtProvider와 같은 클래스를 별도로 만들어서 getRole, getUserId 메서드를 분리해서 구현해주면 구조도 깔끔해지고 JwtAuthenticationFilte 클래스 코드도 더욱 깔끔해질 거 같아요

```java
public JwtParser getJwtParser() {
    return Jwts.parserBuilder()
            .setSigningKey(getSigningKey())
            .build();
}

public Jws<Claims> parseToken(String token) {
    try {
        JwtParser jwtParser = getJwtParser();
        return jwtParser.parseClaimsJws(token);
    } catch (ExpiredJwtException e) {
        // 로그 남기고 custom 예외로 throw
    } catch (UnsupportedJwtException e) {
        // 처리
    }...
}
```

Seung-9 commented on cubco-common/src/main/java/org/cubco/filter/JwtAuthenticationFilter.java:
```diff
@@ -1,4 +1,88 @@
+        ObjectMapper objectMapper = new ObjectMapper();
+        String jsonResponse = objectMapper.writeValueAsString(errorResponse);
+        response.getWriter().write(jsonResponse);
+    }
```
인증 처리 실패 처리는
JwtAuthenticationEntryPoint 클래스로 구현해서 SecurityConfig에 등록해주시면 좋을 거 같아요

그리고 ExceptionHandlerFilter도 구현해주면 필터에서의 전체적인 예외 핸들링이 가능해져서 디버깅에 도움될 것 같습니다!

Seung-9 commented on cubco-common/src/main/java/org/cubco/filter/JwtAuthenticationFilter.java:
```diff
@@ -1,4 +1,88 @@
+        if (token == null || !token.startsWith("Bearer ")) {
+            filterChain.doFilter(request, response);
+            return;
+        }
```
getAccessToken 메서드로 분리해주시면 더욱 깔끔할듯 합니다.

Seung-9 commented on cubco-common/src/main/java/org/cubco/filter/JwtAuthenticationFilter.java:
```diff
@@ -1,4 +1,88 @@
+        }
+
+        try {
+            token = token.substring(7);
```
따로 메서드로 분리하면 가독성 좋을 거 같습니다

Seung-9 commented on cubco-common/src/main/java/org/cubco/util/JWTutil.java:
```diff
@@ -1,4 +1,46 @@
+
+    private SecretKey secretKey;
+
+    public JWTutil(@Value("${jwt.secret}")String secret){
```
이렇게 받아오는 값들이 런타임 이후 변경되지 않기 때문에 보통 final 키워드로 선언해주는데.. @Value를 사용하면 final 키워드를 사용하지 못해서 생성자 바인딩 방식으로 접근하는 것도 좋아보입니다.

hongjunpyo123 commented:
> 고생하셨습니다👍🏻 근데 혹시 JwtUtil을 api모듈이 아닌 common 모듈이 아닌 들어간 이유가 있을까요?

구조에대한 이해가 좀 부족한 것 같습니다.. 모듈별로 무엇이 들어가야 하는지 감이 잘 안잡히는 것도 있구요 의도는 util패키지 자체가 여러 모듈에서 사용될 것 같아서 common쪽에 넣어놨는데 아무래도 몰아넣는 편이 사용하기에 간편해보여 JwtUtil도 넣었습니다

hongjunpyo123 committed:
refactor: JWT claim 설정 방식을 claim()에서 setSubject()로 변경
- .claim(sub, userId) -> .setSubject(String.valueOf(userId))

hongjunpyo123 committed:
refactor: util 패키지를 common에서 API 모듈로 이동

hongjunpyo123 committed:
refactor: 순환 참조 문제 해결 및 의존성 구조 개선
- SecurityConfig, JWTutil 간의 순환 참조 해결
- SecurityFilterChain에서 JWTutil 매개변수로 주입 받도록 변경
- JWTutil 에서 secretKey 생성자 바인딩 방식으로 변경

hongjunpyo123 committed:
refactor: 토큰 추출 로직을 JWTutil 클래스로 분리
- 토큰 추출 메서드 JWTutil 에 추가 extractToken

hongjunpyo123 committed:
refactor: JWT 토큰 만료 시간을 환경변수에서 가져오도록 수정

hongjunpyo123 committed:
refactor: 현재일자, 만료일자 반환 메서드 추가

SeungWoo1222 reviewed (approved):
고생하셨습니다!

Seung-9 committed:
Merge branch 'develop' into feat/#2
